File: frontend/src/features/realtime/schemas.ts
// 백엔드 실시간 이벤트 envelope를 런타임 검증하는 Zod 미러.
import { z } from "zod";

const EnvelopeFields = {
  v: z.literal(1),
  ts: z.number().int(),
};

export const OrderUpdateEnvelopeSchema = z.object({
  ...EnvelopeFields,
  type: z.literal("order_update"),
  payload: z.object({
    order_id: z.string(),
    state: z.string(),
    symbol: z.string(),
    side: z.string(),
    source: z.string(),
  }),
});

export const KillSwitchEnvelopeSchema = z.object({
  ...EnvelopeFields,
  type: z.literal("kill_switch"),
  payload: z.object({
    event_id: z.string(),
    trigger_type: z.string(),
  }),
});

export const KillSwitchResolvedEnvelopeSchema = z.object({
  ...EnvelopeFields,
  type: z.literal("kill_switch_resolved"),
  payload: z.object({
    event_id: z.string(),
    trigger_type: z.string(),
  }),
});

export const SessionStateEnvelopeSchema = z.object({
  ...EnvelopeFields,
  type: z.literal("session_state"),
  payload: z.object({
    session_id: z.string(),
  }),
});

export const RealtimeEnvelopeSchema = z.discriminatedUnion("type", [
  OrderUpdateEnvelopeSchema,
  KillSwitchEnvelopeSchema,
  KillSwitchResolvedEnvelopeSchema,
  SessionStateEnvelopeSchema,
]);

export type RealtimeEnvelope = z.infer<typeof RealtimeEnvelopeSchema>;

export function parseRealtimeEnvelope(value: unknown): RealtimeEnvelope | null {
  if (typeof value !== "string") return null;
  try {
    const raw: unknown = JSON.parse(value);
    const result = RealtimeEnvelopeSchema.safeParse(raw);
    return result.success ? result.data : null;
  } catch {
    return null;
  }
}

File: backend/src/realtime/schemas.py
# 실시간 WebSocket 이벤트와 인증 메시지 계약의 단일 정의.
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict

USER_CHANNEL_PREFIX = "qb:rt:user:"
WS_CLOSE_AUTH_FAILED = 4401
WS_CLOSE_ORIGIN_DENIED = 4403
WS_CLOSE_CONNECTION_LIMIT = 4408


class RealtimeEnvelope(BaseModel):
    """Redis pubsub와 WebSocket이 공유하는 이벤트 envelope."""

    v: int = 1
    type: Literal["order_update", "kill_switch", "kill_switch_resolved", "session_state"]
    ts: int
    payload: dict[str, object]


class OrderUpdatePayload(BaseModel):
    """주문 상태 변경 이벤트 payload."""

    model_config = ConfigDict(extra="ignore")

    order_id: str
    state: str
    symbol: str
    side: str
    source: str


class KillSwitchPayload(BaseModel):
    """킬 스위치 변경 이벤트 payload."""

    model_config = ConfigDict(extra="ignore")

    event_id: str
    trigger_type: str


class SessionStatePayload(BaseModel):
    """트레이딩 세션 상태 변경 이벤트 payload."""

    model_config = ConfigDict(extra="ignore")

    session_id: str


PAYLOAD_MODELS: dict[str, type[BaseModel]] = {
    "order_update": OrderUpdatePayload,
    "kill_switch": KillSwitchPayload,
    "kill_switch_resolved": KillSwitchPayload,
    "session_state": SessionStatePayload,
}


class AuthMessage(BaseModel):
    """WebSocket 최초 인증 메시지."""

    type: Literal["auth"]
    token: str


def user_channel(user_id: str) -> str:
    """채널 키는 users.id 내부 UUID 문자열이며 clerk_user_id가 아니다."""
    return f"{USER_CHANNEL_PREFIX}{user_id}"
